dlt/common/libs/pyarrow.py

```diff
@@ -634,124 +649,151 @@ def concat_batches_and_tables_in_order(
+
+    # case 2: if `base case` failed, the array contains non-string objects (likely dict, tuple, list, set, etc.)
+    # encode these objects as JSON strings; 
+    if arrow_array is None and dlt_data_type == "json":
+        logger.warning(
+            "Received `data_type='json'`, data requries serialization to string, slowing extraction. "
+            "Cast the JSON field to STRING in your database system to improve performance. "
+            "For example, create and extract data from an SQL VIEW that SELECT with CAST."
+        )
+        arrow_array = pa.array(
+            [
+                (
+                    None if item is None
+                    else json_to_str(list(item)) if issubclass(type(item), set)  # sets are not supported by json_to_str
+                    else json_to_str(item)
                 )
-
-        # New schema
-        column_order = {name: idx for idx, name in enumerate(columns)}
-        arrow_schema = pa.schema(
-            sorted(
-                list(arrow_schema) + new_schema_fields,
-                key=lambda x: column_order[x.name],
-            )
+                for item in column_data
+            ],
+            type=pa.string()
         )
 
-    return pa.Table.from_pydict(columnar_known_types, schema=arrow_schema)
+    # case 3: backends store decimals in various format (binary, bytes, etc.) and returned Python values have inconsistent types
+    # `base case` fails because the Python type received is incompatible with the inferred pyarrow type (e.g., `pyarrow.decimal128()`) 
+    if arrow_array is None and dlt_data_type in ("decimal", "wei"):
+        # let pyarrow infer type (e.g., float, string) THEN cast it to the inferred type
+        tmp_array = pa.array(column_data)
+        try:
+            arrow_array = tmp_array.cast(inferred_arrow_type)
+        except pa.ArrowInvalid:
+            pass
+
+    if arrow_array is None:
+        raise PyToArrowConversionException()
+
+    return arrow_array
+
+
+def row_tuples_to_arrow(
+    rows: TDataItems, caps: DestinationCapabilitiesContext, columns: TTableSchemaColumns, tz: str
+) -> Any:  # pyarrow.Table
+    """Converts the rows to an arrow table using the columns schema.
+    1. Pivot rows into columns.
+    2. Convert columns to pyarrow arrays; coerce type and nullability; exclude types unsupported by arrow 
+    3. Create table
+    4. Remove columns full of null values
+    """
+    from dlt.common.libs.pyarrow import pyarrow as pa
+
+    columnar = transpose_rows_to_columns(rows, column_names=columns.keys())
+
+    arrow_arrays = []
+    arrow_fields = []
+    for column_name, column_data in columnar.items():
+        column_schema = columns[column_name]
+
+        try:
+            arrow_array = convert_numpy_to_arrow(column_data, caps, column_schema, tz)
+        # TODO if converting to arrow fail, should we raise or skip column?
+        except PyToArrowConversionException as e:
```

Comment:
we should definitely fail. what could be cool though if the exception we raise always has information about column name. maybe we can reuse the exception you worked on recently?

in that case user can remove data and try again